Loop/src/loop.hxx

```diff
@@ -517,6 +521,110 @@ public:
+  // from those that CarpetX identifies as boundary points.
+  // This excludes ghost faces, but includes ghost edges/corners on non-ghost faces.
+  // Loop over faces first, then edges, then corners.
+  template <int CI, int CJ, int CK, int VS = 1, int N = 1, typename F>
+  inline CCTK_ATTRIBUTE_ALWAYS_INLINE void
+  loop_radbnd(const vect<int, dim> &group_nghostzones, const F &f) const {
+    // boundary_box sets bnd_min and bnd_max
+    vect<int, dim> bnd_min, bnd_max;
+    // if on (Carpetx) boundary points, then
+    //   bnd_min = nghostzones - lbnd
+    //   bnd_max = gsh - offset - nghostzones - lbnd
+    //   where lbnd : an array of cctk dim integers containing the lowest index
+    //   (in each direction) of the local grid, as seen on the global grid
+    // else,
+    //   bnd_min = -100000000
+    //   bnd_max = gsh + 1000000
+    //   where cctk_gsh : an array of cctk dim integers with the global grid size
+    // see line 190
+    boundary_box<CI, CJ, CK>(group_nghostzones, bnd_min, bnd_max);
+
+    // domain_boxes sets all_min, all_max, int_min and int_max
+    vect<int, dim> all_min, all_max, int_min, int_max;
+    // Indices for the entire grid
+    //   all_min = ghost_offset;
+    //   all_max = lsh - offset - ghost_offset;
+    // Indices for the interior
+    //   int_min = nghostzones;
+    //   int_max = lsh - offset - nghostzones;
+    // see line 203
+    domain_boxes<CI, CJ, CK>(group_nghostzones, all_min, all_max, int_min,
+                             int_max);
+
+    // rank selects: faces, edges, corners
+    // rank(interior) = 3
+    // rank(face)     = 2
+    // rank(edge)     = 1
+    // rank(corner)   = 0
+    for (int rank = dim - 1; rank >= 0; --rank) {
+
+      // Nested loops that determine {ni,nj,nk} components of the normal vector
+      for (int nk = -1; nk <= +1; ++nk) {
+        for (int nj = -1; nj <= +1; ++nj) {
+          for (int ni = -1; ni <= +1; ++ni) {
+            // True when the normal vector is normal to a {face, edge, corner}
+            if ((ni == 0) + (nj == 0) + (nk == 0) == rank) {
+              // True when point is on left/right boundary,
+              // and vector is not parallel to a {face,corner,edge}
+              // In either of the 3 directions
+              if ((ni != 0 && bbox[ni < 0 ? 0 : 1][0]) ||
+                  (nj != 0 && bbox[nj < 0 ? 0 : 1][1]) ||
+                  (nk != 0 && bbox[nk < 0 ? 0 : 1][2])) {
+
+                const vect<int, dim> inormal{ni, nj, nk};  //define normal vector
+
+                vect<int, dim> imin, imax;
+                for (int d = 0; d < dim; ++d) { // Set imin, imax in each direction
+                  switch (inormal[d]) {
+                  case -1: // lower boundary
+                    // shift "upwards" from physical boundary by # of ghostpoints
+                    //                                      // xxx|~~~~~~~~~~~~~~|~~~
+                    imin[d] = all_min[d] + nghostzones[d];  // ~~~ ^~~~~~~~~~~~~~ ~~~
```

Comment:
It seems that `all_min` and `int_min` are always shifted upwards, and `all_max` and `int_max` are always shifted downwards. Instead of using this `for` loop with the `switch` logic, you could just write `imin = all_min + nghostzones` etc., similar to the way `bnd_min` is handled below.

Reply:
That is a good observation. I updated it to apply the upward shifts to `all_min` and `int_min`, and downward shifts to `all_max` and `int_max` all in one place. However, it seems to me that the switch is still necessary, depending on whether the point lies on the upper/lower boundary or interior. If there is another way to do it without the switch, could you spell it out more explicitly for me?